### DISCUSSION TITLE
icharger 4880 easun 80a

### DISCUSSION THREAD

**Discussion by lucasimons** [syssi/esphome-jnge-mppt-controller#36] [General]
Hi @syssi 

I found this communication protocol file there is no possibility to adapt the jnge code for this very popular controller on aliexpress, 
icharger 4880 80a easun powland must.

thanks for all

[MPPT_4880_PC1800_RS485_Modbud_RTU_communication_Protocol1_4_14_MAP&nbsp;.xlsx](https://github.com/syssi/esphome-jnge-mppt-controller/files/10492617/MPPT_4880_PC1800_RS485_Modbud_RTU_communication_Protocol1_4_14_MAP.nbsp.xlsx)

![image](https://user-images.githubusercontent.com/67467751/214369917-cbcb4b61-ff61-4334-833f-6b88bc293721.png)

![image](https://user-images.githubusercontent.com/67467751/214369978-3a9cb7b4-cefd-47de-ae7e-7a05acf1d71e.png)

**Comment by syssi** [comment]
You could use the `modbus_controller` component to talk to the inverter. @SeByDocKy is this the same inverter you implemented recently using the `modbus_controller`?

See https://esphome.io/components/modbus_controller.html

**Reply by lucasimons** [reply]
This is a separate solar charge controller not a inverter, so I can use the modbus controller can you explain how I can set the write command for example the address 10120 to set the max charging current?

**Reply by SeByDocKy** [reply]
> You could use the `modbus_controller` component to talk to the inverter. @SeByDocKy is this the same inverter you implemented recently using the `modbus_controller`?
> 
> See https://esphome.io/components/modbus_controller.html

No was two solar charger (Epever & SRNE) not an inverter.... but I don't think it's super different.... If you have the protocol, then all registers.... can be done in a couple of hours with builtin modbus fonctions of ESPhome

Btw, we did a small PCB 

![1673366373346](https://user-images.githubusercontent.com/6464595/214389788-aca62caf-ca91-427c-8098-70312c5510f0.jpg)

**Reply by lucasimons** [reply]
@syssi @SeByDocKy  


Thanks for the advice, I still need to figure out how to write to the registry to change the parameters, do you have any advice?


![image](https://user-images.githubusercontent.com/67467751/214602829-2720b843-1768-40b8-aeb7-a25f7b43061d.png)


```


number: 
   
  - platform: modbus_controller
    id: mppt_max_current
    name: MPPT Corrente Max 
    address: 10108
    value_type: U_WORD
    register_type: holding
    unit_of_measurement: "A"
    icon: mdi:timer-sand
    min_value: 1
    max_value: 80
    step: 1
    lambda: "return  x * 0.1; "
    write_lambda: |-
      ESP_LOGD("main","Modbus Number incoming value = %f",x);
      uint16_t b_capacity = x  ;
      payload.push_back(b_capacity);
      return x  ;


```

**Reply by syssi** [reply]
Do I understand your screenshot correctly? You are able to receive sensor values now using the `modbus_controller` component? Could you provide this configuration for the community to make it easier for others to talk to this device?

I will try to get an idea how to pass settings to the device using the `modbus_controller`.

**Reply by SeByDocKy** [reply]
> @syssi @SeByDocKy
> 
> Thanks for the advice, I still need to figure out how to write to the registry to change the parameters, do you have any advice?
> 
> ![image](https://user-images.githubusercontent.com/67467751/214602829-2720b843-1768-40b8-aeb7-a25f7b43061d.png)
> 
> ```
> 
> 
> number: 
>    
>   - platform: modbus_controller
>     id: mppt_max_current
>     name: MPPT Corrente Max 
>     address: 10108
>     value_type: U_WORD
>     register_type: holding
>     unit_of_measurement: "A"
>     icon: mdi:timer-sand
>     min_value: 1
>     max_value: 80
>     step: 1
>     lambda: "return  x * 0.1; "
>     write_lambda: |-
>       ESP_LOGD("main","Modbus Number incoming value = %f",x);
>       uint16_t b_capacity = x  ;
>       payload.push_back(b_capacity);
>       return x  ;
> ```




I think a modbus number or select is way to go to write parameters. I am doing like this in general. Be aware, for some solar charger, you need to write several registers at the same time. You can have a look to my Epever ESPhome integrations

**Reply by lucasimons** [reply]
> Do I understand your screenshot correctly? You are able to receive sensor values now using the `modbus_controller` component? Could you provide this configuration for the community to make it easier for others to talk to this device?
> 
> I will try to get an idea how to pass settings to the device using the `modbus_controller`.

yesterday the sensor works fine but this morning sometime sho wrong value 
![image](https://user-images.githubusercontent.com/67467751/214782843-1606c1bc-d66a-4dcc-a4b5-66188ef28f68.png)



the code is [here](https://github.com/lucasimons/MPPT-4880/commit/bc902280a805cebaa2ab6eece96445896fd0453e)

**Comment by SeByDocKy** [comment]
Probably you need to multiply par 0.01  and 100 respectively between sensor and number.

**Reply by lucasimons** [reply]
I did various tests and the sensor works but I can't change the values, current or voltage


```
  - platform: modbus_controller
    id: mppt_max_current
    name: MPPT Corrente Max 
    address: 10108
    value_type: U_WORD
    register_type: holding
    unit_of_measurement: "A"
    icon: mdi:timer-sand
    min_value: 1
    max_value: 80
    step: 1
    lambda: "return  x * 0.1; "
    write_lambda: |-
      ESP_LOGD("main","Modbus Number incoming value = %f",x);
      uint16_t b_capacity = x*10  ;
      payload.push_back(b_capacity);
      return x  ;

```



the wrong value i think is a problem of interferences because i use an esp32 module with i read the value of bms mppt victron and mppt easun, the bms with ant-bms integration (by @syssi) and it works fine but the mppts easun and victron there are some error






```
[14:58:33][W][victron:093]: Last transmission too long ago
[14:58:35][D][modbus_controller:032]: Modbus command to device=1 register=0x2778 countdown=0 no response received - removed from send queue
[14:58:36][W][victron:093]: Last transmission too long ago
```

**Reply by syssi** [reply]
Please enable the `debug` mode of the UART component and provide the raw data of the TX (`>>>`) & RX (`<<<`) frames.

**Reply by lucasimons** [reply]
I had deactivated it because the sensors work but now I have these strange interferences

this is the log with debug enable

```
[15:09:21][D][uart_debug:114]: >>> 01:03:3B:61:00:03:59:31
[15:09:21][W][modbus:105]: Modbus CRC Check failed! 77DE!=301
[15:09:21][W][modbus:105]: Modbus CRC Check failed! 8D2E!=1A01
[15:09:21][W][modbus:105]: Modbus CRC Check failed! B3BE!=202
[15:09:21][W][modbus:105]: Modbus CRC Check failed! 7A82!=201
[15:09:22][D][uart_debug:114]: <<< 02:00:02:78:B4:01:03:00:01:1A:38:1F:01:03:02:01:1A:30:1F:00:03:02:01:1A:38:1F:01:03:02:01:1A:38:1F:01:03:02:01:1A:30:1F:01:02:02:02:58:B8:9C:01:02:00:02:58:B8:DE:01:03:02:02:58:B8:DE:01:03:02:02:58:B8:9C:01:03:02:02:58:B8:DE:01:03:06:00:02:00:00:00:02:78:B4:01:03:06:00:02:00:02:00:02:78:B4:01:03:06:00:02:00:02:00:02:78:B4:01:03:06:00:02:00:02:00:02:78:B4:01

```

[logs_sottoscala-esp32-2_upload.txt](https://github.com/syssi/esphome-jnge-mppt-controller/files/10510022/logs_sottoscala-esp32-2_upload.txt)

**Reply by lucasimons** [reply]
[logs_sottoscala-esp32-2_logs.txt](https://github.com/syssi/esphome-jnge-mppt-controller/files/10510416/logs_sottoscala-esp32-2_logs.txt)
now the sun is very low and all sensor works fine but i can t cange any value

![image](https://user-images.githubusercontent.com/67467751/214866435-0072c5f9-3167-4d25-8b63-517ecd7c53bd.png)


in this file i has tried to edit the maximum charge current 60A to lower value

**Reply by lucasimons** [reply]
Now i use two different esp32, one with i read the value of victron mppt and the second esp32 for the easun mppt but there are some interferences they are connected with the same gnd because i use a dc dc converter 24 to 12 (orion victron), but sometimes they work fine today untill 9:00 am 
![image](https://user-images.githubusercontent.com/67467751/215444242-3510845c-91ae-4291-8f20-7dc2eb8b2199.png)

 

![image](https://user-images.githubusercontent.com/67467751/215443350-f8a122a8-d54c-48f3-81aa-fa8774ad0f62.png)


https://youtu.be/Bl_VvPQo2-s

**Comment by grydo** [comment]
Hi @syssi @SeByDocKy
Has anyone managed to integrate it?

**Reply by syssi** [reply]
If you are still looking for a solution: https://github.com/lucasimons/MPPT-4880
